### PR TITLE
Implementing Writeable for all integers

### PR DESCRIPTION
**Pull request by robertbastian**

### CODE REVIEW

File: utils/writeable/src/impls.rs [hunk 1]
```diff
@@ -8,84 +8,82 @@ use core::convert::TryFrom;
...
+            }
 
-    fn write_len(&self) -> LengthHint {
-        if *self < 10 {
-            LengthHint::exact(1)
-        } else if *self < 100 {
-            LengthHint::exact(2)
-        } else {
-            LengthHint::exact(3)
+            fn write_len(&self) -> LengthHint {
```
Comment (sffc) [inline utils/writeable/src/impls.rs:38]: Thought: I am aware of at least 3 ways to compute this.

1. Divide by 10 in a loop, as you are doing
2. Lookup table, similar to what we had before (this can be a binary search to speed it up)
3. Computing `self.log10() + 1`

I'd be interested to see both performance and code size impacts of those three approaches.

If you don't want to do this now, please file an issue for someone to take in the future.

**Review by sffc — APPROVED** [review]

File: utils/writeable/src/impls.rs (hunk 1 above)
Comment (robertbastian) [reply utils/writeable/src/impls.rs:38]: I don't think we should use `.log10()`. If there's no hardware floating point unit it will probably run some Newton approximation in ints to an accuracy we don't need which will get pretty expensive.

We can take a cheap floor-log2 by inspection the highest bit, then we're off by at most one digit, which we can fix with one `pow` invocation.

Comment (robertbastian) [reply utils/writeable/src/impls.rs:38]: In my rough benchmarking (0..25_000_000) I see
1. Divide by 10 in a loop: 1.9s
2. Binary search over hardcoded array: 3.8s (+ probably biggest code size)
3. `log10`: 1.8s on `x86_64`, but also `no_std` doesn't seem to include logarithms
4. `log10` approximation: 1.2s

File: utils/writeable/src/impls.rs [hunk 2]
```diff
@@ -8,84 +8,103 @@ use core::convert::TryFrom;
...
+    assert_writeable_eq!(&10u128, "10");
+    assert_writeable_eq!(&10i128, "10");
+    assert_writeable_eq!(&-10i128, "-10");
+    assert_writeable_eq!(&99u128, "99");
+    assert_writeable_eq!(&99i128, "99");
+    assert_writeable_eq!(&-99i128, "-99");
+    assert_writeable_eq!(&u128::MAX, u128::MAX.to_string());
+    assert_writeable_eq!(&i128::MAX, i128::MAX.to_string());
+    assert_writeable_eq!(&i128::MIN, i128::MIN.to_string());
 }
```
Comment (sffc) [inline utils/writeable/src/impls.rs:103]: Nit: Add more tests.  Suggestions:

- More tests around power-of-10 boundaries (99, 100, 999, 1000, ...)
- Randomly generated integers of each size (u8, u16, u32, u64, u128) that check the Writeable impl against the ToString impl

This can be done in a follow-up PR if you need to merge this one for a dependency.

**Review by sffc — APPROVED** [review]
Praise: Very nice and clever code!
